news-explorer-frontend/src/components/PopupWithForm/PopupWithForm.js

```diff
@@ -0,0 +1,310 @@
+    const email = emailInputfield.value;
+    const password = passwordInputfield.value;
+    const username = nameInputfield.value;
+
+    if (isSignInPopupOpen) {
+      handleSigninSubmit(email, password);
+    }
+
+    if (isRegisterPopupOpen) {
+      handleSignUpSubmit(email, password, username);
+    }
+  };
+
+  const switchModelHandler = () => {
+    setSubmitButtonClass('popup__submit-btn_disabled');
+    setFormIsValid(false);
+    props.setIsSignInPopupOpen((prevState) => !prevState);
+    props.setIsRegisterPopupOpen((prevState) => !prevState);
+    setEmailInputField({
+      value: '',
+    });
+    setNameInputField({
+      value: '',
+    });
+    setPasswordInputField({
+      value: '',
+    });
+  };
+
+  const handleEmailValidation = (field, value) => {
+    if (!value.match(/^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/)) {
+      setEmailInputField({
+        field,
+        value,
+        error: 'not valid email',
+        isValid: false,
+      });
+    }
+    else {
+      setEmailInputField({
+        field,
+        value,
+        error: '',
+        isValid: true,
+      });
+    }
+  };
+
+  const handleNameValidation = (field, value) => {
+    if (!value.match(/^[a-zA-Z]+$/)) {
+      setNameInputField({
+        field,
+        value,
+        error: 'not valid name',
+        isValid: false,
+      });
+    }
+    else {
+      setNameInputField({
+        field,
+        value,
+        error: '',
+        isValid: true,
+      });
+    }
+  };
+
+  const handlePasswordValidation = (field, value) => {
+    if (value.length < 2) {
+      setPasswordInputField({
+        field,
+        value,
+        error: 'minimum 2 length needed',
+        isValid: false,
+      });
+    }
+    else {
+      setPasswordInputField({
+        field,
+        value,
+        error: '',
+        isValid: true,
+      });
+    }
+  };
+
+  const handleValidation = (field, value) => {
+    if (field === 'email') {
+      handleEmailValidation(field, value);
+    }
+    if (field === 'name') {
+      handleNameValidation(field, value);
+    }
+    if (field === 'password') {
+      handlePasswordValidation(field, value);
+    }
+  };
+
+  const handleChange = (e) => {
+    const field = e.target.name;
+    const value = e.target.value;
+    setFormErrorMessage('');
+    handleValidation(field, value);
+  };
+
+  return (
+    <div className={`popup ${formIsOpen}`} onClick={handleClickOnOverlayClose} tabIndex='0'>
+      <div className='popup__content' onClick={(e) => e.stopPropagation()}>
+
+        <button className={`popup__close-btn popup__close-btn_${props.name}`} onClick={close}>
+          <img src={closebtn} alt='close-btn' className='popup__close-btn-img' />
+        </button>
+
+        <form onSubmit={handleFormSubmit} className='popup__form'>
+          <h2 className='popup__title'>
+            {isSignInPopupOpen ? 'Sign in' : 'Sign up'}
+          </h2>
+          <label htmlFor='email' className='popup__label'>email</label>
```

Comment:
`PopupWithForm` should be a universal component. The inputs should be passed as `children` because any form can have different number of inputs